### PR TITLE
Implement lightning.kachow!

### PR DESCRIPTION
**Pull request by JayOrten**
This PR implements PyTorch Lightning functionality

The major contributions of this PR are:
- train_model_lightning.py: implements a training script utilizing PyTorch Lightning
- dataset.py: contains the PyTorch Lightning DataModule class
- models.py: contains the Torch Lightning module model classes and the HuggingFace module model classes
- generation.py: generation script for generating text with a trained model
- inference.py: inference script for running trained model on test set.
- Adds additional parameters to configuration file
- Cleans up configuration and slurm files

### CODE REVIEW

File: TODO.md
Comment (nprisbrey) [inline TODO.md:1]: I assume we're working towards deleting this file, correct?

File: configs/template_config.yaml [hunk 1]
```diff
@@ -23,7 +23,7 @@ dataset_feature: "text"
 # HuggingFace Dataset Name (str)
 dataset_name: "wikitext"
 # HuggingFace Dataset Subset (str)
-dataset_subset: "wikitext-2-v1"
+dataset_subset: "wikitext-103-raw-v1"
```
Comment (nprisbrey) [inline configs/template_config.yaml]: Is there a reason that we're changing from the lightest dataset to get someone up and running or is this just residual code?

File: slurm/slurm_template.sh [hunk 2]
```diff
@@ -1,41 +1,17 @@
...
-#SBATCH --qos=dw87
-#SBATCH -J "template_name"   # job name
+#SBATCH --gres=gpu:1
+#SBATCH --qos=cs
+#SBATCH --mem=64G   # memory per CPU core
+#SBATCH -J "train"   # job name
 #SBATCH --output=%x_%j.out
 
 # Set the max number of threads to use for programs using OpenMP. Should be <= ppn. Does nothing if the program doesn't use OpenMP.
 export OMP_NUM_THREADS=$SLURM_CPUS_ON_NODE
```
Comment (nprisbrey) [inline slurm/slurm_template.sh]: With our new changes of moving to different QOS and changing our requirements, does it still make sense to be exporting this OpenMP variable?

File: configs/template_config.yaml [hunk 3]
```diff
@@ -44,12 +44,41 @@ models_path: "/YOUR/PATH/HERE/data/models"
...
+save_top_k: 3
+# val_check_interval (float): How often to validate as a fraction of an epoch.
+val_check_interval: 0.5
+# use_slurm (bool): Whether to use SLURM for training.
+use_slurm: true
+# gamma (float): Learning rate scheduler gamma.
+gamma: 0.85
+# accumulate_grad_batches (int): Accumulate gradients over n batches.
+accumulate_grad_batches: 1
+
```
Comment (nprisbrey) [inline configs/template_config.yaml]: Is there a reason that we're pulling this out of the "MODEL TRAINING CONFIGURATION" section below?

File: configs/template_config.yaml [hunk 4]
```diff
@@ -44,12 +44,41 @@ models_path: "/YOUR/PATH/HERE/data/models"
 # TensorBoard Path
 tboard_path: ~
 
+# Checkpoint Path: path to checkpoint to load for generation/inference
+checkpoint_path: ~
```
Comment (nprisbrey) [inline configs/template_config.yaml]: If we are to implement a `checkpoint_path` argument, shouldn't this be implemented everywhere (ex: the normal `train_model.py`)?

File: train_model_lightning.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    with open(config_path, 'r') as f:
+        config = yaml.safe_load(f)
+        
+    config = Struct(**config)
+
+    start_time = time.time()
+
+    model, test_loss = train_model(config)
+
+    end_time = time.time()
```
Comment (nprisbrey) [inline train_model_lightning.py]: What is the point of these times? It doesn't seem that they're used anywhere....

File: train_model_lightning.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    # Create unique label for model (model type, parameter count, hyperparameters**)
+    model_label = f"{config.model_type}_{total_params}_LR{config.learning_rate}_ED{config.embed_dim}_FFN{config.ffn_dim}_H{config.heads}_S{config.seq_len}_{datetime.now().strftime('%Y-%m-%d-%H:%M:%S')}"
+
+    # Initialize model directory for config files, weights, etc.
+    model_dir = Path(config.models_path) / model_label
+    model_dir.mkdir(parents=True, exist_ok=False)
+    print(f"Saving model files in {model_dir}")
+
+    # Initialize checkpoints directory
+    checkpoints_dir = model_dir / "checkpoints"
```
Comment (nprisbrey) [inline train_model_lightning.py]: Seems like this should be refactored to use the new config parameter.

**Review by nprisbrey — COMMENTED** [review]
Some comments on the current status of things. Let it be recognized that I may have just been commenting on work that was already planning on being done but hasn't come to pass yet.

File: TODO.md
Comment (JayOrten) [reply TODO.md:1]: Yes

File: configs/template_config.yaml (hunk 1 above)
Comment (JayOrten) [reply configs/template_config.yaml]: Nope, just residual. Changed

File: slurm/slurm_template.sh (hunk 2 above)
Comment (JayOrten) [reply slurm/slurm_template.sh]: Not necessarily, we likely weren't using OpenMP anyways for anything, QOS wouldn't change that. This is here simply because ORC recommends it, but we could remove it.

File: configs/template_config.yaml (hunk 3 above)
Comment (JayOrten) [reply configs/template_config.yaml]: I would argue yes. Model Training Configuration should have the parameters for actual model configuration, i.e., layers, width, etc. These parameters are settings for the devices and the hyperparameters, such as LR, epochs, etc. I would say separating them makes it easier to parse through. These also need to be reorganized a bit.

File: configs/template_config.yaml (hunk 4 above)
Comment (JayOrten) [reply configs/template_config.yaml]: The purpose of this parameter is that, once you've trained a model and want to do inference or generation, you put into this field the path to the checkpoint that you want to use for those tasks. This isn't used in the training scripts.

File: train_model_lightning.py (hunk 5 above)
Comment (JayOrten) [reply train_model_lightning.py]: removed

File: train_model_lightning.py (hunk 6 above)
Comment (JayOrten) [reply train_model_lightning.py]: See other comment on config parameter

File: slurm/slurm_template.sh (hunk 2 above)
Comment (JayOrten) [reply slurm/slurm_template.sh]: Update on this: removing this line can result in strange behaviour.

**Comment by JayOrten** [comment]
Based on the state of the code and results from test runs I feel comfortable opening this up

File: train_model_lightning.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import sys
+import yaml
+
+# Torch
+import torch
+from torchinfo import summary as model_summary
+
+# PyTorch Lightning
+from pytorch_lightning import Trainer
+from pytorch_lightning.plugins.environments import SLURMEnvironment
```
Comment (KimballNJardine) [inline train_model_lightning.py]: Alphabetize

File: configs/template_config.yaml [hunk 8]
```diff
@@ -59,35 +92,20 @@ activation_dropout: 0.0
...
 layers: 2
-# Learning Rate (float)
-learning_rate: 0.001
 # Model Type (str): Name of model architecture to train
-model_type: "retnet" # Choices: "retnet", "transformer"
-# Number of GPUs (int)
-num_devices: 1
-# Random Seed (int): For reproducibility
-rand_seed: 42
+model_type: retnet # Choices: "retnet", "transformer"
```
Comment (nprisbrey) [inline configs/template_config.yaml]: Is there a reason why we're making strings no longer in double quotes (ex: `"retnet"`)?

File: dataset.py
Comment (nprisbrey) [inline dataset.py:1]: It seems that most typing hints and all documentation strings are missing from this file.

File: dataset.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+from pathlib import Path
+from pytorch_lightning import LightningDataModule
+from typing import List, Optional
+from torch.utils.data import DataLoader
+
+from datasets import load_dataset
+
+class DataModule(LightningDataModule):
+    def __init__(self, config, num_workers=0):
```
Comment (nprisbrey) [inline dataset.py]: What's stopping us from putting `num_workers` into the config file like everything else? This could also maintain consistency with `train_model_lightning.py` and remove a constant from both files.

File: slurm/train_model.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+#SBATCH --requeue
+#SBATCH --signal=SIGHUP@90
+#SBATCH --output=%x_%j.out
+
+# Set the max number of threads to use for programs using OpenMP. Should be <= ppn. Does nothing if the program doesn't use OpenMP.
+export OMP_NUM_THREADS=$SLURM_CPUS_ON_NODE
+
+# LOAD MODULES, INSERT CODE, AND RUN YOUR PROGRAMS HERE
+mamba activate retnet
+srun python3 ../../train_model_lightning.py ../../configs/user_configs/YOUR_CONFIG_HERE.yaml
```
Comment (nprisbrey) [inline slurm/train_model.sh]: Just double checking: is calling `srun` here for a job step with Slurm done on purpose for the requeueing logic or is this extra and could be removed?

File: models.py
Comment (nprisbrey) [inline models.py:1]: Another reminder of updating Doc Strings.

File: train_model_lightning.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+# Local
+from models import RetNetModel, TransformerModel
+from dataset import DataModule
+from utils import Struct
+
+# Allow torch to run float32 matrix multiplications in lower precision for
+# better performance while training if hardware is capable
+torch.backends.cuda.matmul.allow_tf32 = True
+
+class CustomCheckpoint(ModelCheckpoint):
```
Comment (nprisbrey) [inline train_model_lightning.py]: Can we rename this possibly to `CustomModelCheckpoint` or something similar to be more descriptive?

File: configs/template_config.yaml [hunk 12]
```diff
@@ -59,35 +92,20 @@ activation_dropout: 0.0
...
 dropout: 0.1
 # Embedding Dimension (int): Embedding dimension size of each token
 embed_dim: 80
-# Epochs (int): Number of epochs to train for
-epochs: 1
 # FFN Dimension (int): Hidden layer size of Feed Forward Network (FFN)
 ffn_dim: 12
-# Fully Sharded Data Parallel (bool): Whether to shard Module parameters across
-# data parallel workers or not (with the FairScale library)
-fsdp: false
```
Comment (nprisbrey) [inline configs/template_config.yaml:74]: The normal `train_model.py` script no longer runs because `fsdp` and `validation_frequency` are no longer included in the config file. How are we expecting to resolve this?

File: slurm/train_model.sh
Comment (nprisbrey) [inline slurm/train_model.sh:1]: Do we have a reasoning for why we only have a `train_model_lightning.py` script in the `slurm` directory and only have a `train_model.py` script in the `scripts` directory? `inference.sh` and `generate.sh` don't seem to have any equivalent in the `scripts` directory either.

**Review by nprisbrey — CHANGES_REQUESTED** [review]
These are some preliminary thoughts @JayOrten. Let me know what you think and then I'll finish looking over the code. I already made some changes myself but would also like to see what your thoughts are on changing more.

File: configs/template_config.yaml (hunk 8 above)
Comment (JayOrten) [reply configs/template_config.yaml]: Technically quotes are unnecessary in yaml, but changed for consistency

File: slurm/train_model.sh (hunk 10 above)
Comment (JayOrten) [reply slurm/train_model.sh]: Yes, without `srun` the script will hang while initializing devices

File: configs/template_config.yaml (hunk 12 above)
Comment (JayOrten) [reply configs/template_config.yaml:74]: I will add them back in. However, we should consider whether or not we want to maintain both `train_model.py` and `train_model_lightning.py`, because the latter should be able to do everything the former is doing, while enabling multi-device training.

File: dataset.py (hunk 9 above)
Comment (JayOrten) [reply dataset.py]: Added. There wasn't any reason why, I just haven't experimented with that parameter as much

File: train_model_lightning.py (hunk 11 above)
Comment (JayOrten) [reply train_model_lightning.py]: changed

File: slurm/train_model.sh
Comment (JayOrten) [reply slurm/train_model.sh:1]: Nope, I've just only been testing through slurm. Added

File: models.py
Comment (JayOrten) [reply models.py:1]: added, let me know if I missed anything that should be explained more fully

File: configs/template_config.yaml (hunk 12 above)
Comment (nprisbrey) [reply configs/template_config.yaml:74]: I think that, as you mention, we should just drop the original `train_model.py` script in favor of Lightning. Thoughts @JayOrten and @KimballNJardine?

Comment (JayOrten) [reply configs/template_config.yaml:74]: I'm okay with that

Comment (KimballNJardine) [reply configs/template_config.yaml:74]: That is fine with me.

**Review by nprisbrey — APPROVED** [review]
Looks good! All scripts in the `scripts` directory function as expected. I haven't tested any of the Slurm scripts myself, though they run on the same underlying code and I believe they should be easy to use for this weekend.

**Review by KimballNJardine — APPROVED** [review]
Looks good. Approved.
